### PR TITLE
[WIP] Feature/clandmark preset

### PR DESCRIPTION
**Pull request by lloydmeta**
Would like some feedback on this PR. In short: this is me hacking things together until:

- `./cppbuild.sh install clandmark`
- `mvn clean install --projects clandmark -X`

both ran without errors.

### CODE REVIEW

File: clandmark/cppbuild.sh [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+        ;;
+    linux-armhf)
+        CC=arm-linux-gnueabihf-gcc CXX=arm-linux-gnueabihf-g++ $CMAKE -DCMAKE_BUILD_TYPE=Release -DOpenCV_DIR=$OPENCV_PATH/share/OpenCV/ -DCMAKE_INSTALL_PREFIX=..
+        make -j4
+        make install clandmark
+        ;;
+    macosx-*)
+        CXX="g++ -fpermissive" $CMAKE -DCMAKE_BUILD_TYPE=Release -DOpenCV_DIR=$OPENCV_PATH/share/OpenCV/ -DCMAKE_INSTALL_PREFIX=..
+        make -j4
+        make install/strip
```
Comment (lloydmeta) [inline clandmark/cppbuild.sh]: I've only tested this one, since I only have 1 OS

File: clandmark/pom.xml [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+    <parent>
+        <groupId>org.bytedeco</groupId>
+        <artifactId>javacpp-presets</artifactId>
+        <version>1.2-SNAPSHOT</version>
+    </parent>
+
+    <groupId>org.bytedeco.javacpp-presets</groupId>
+    <artifactId>clandmark</artifactId>
+    <!-- TODO figure out what version CLandmark is actually at?!?!? -->
+    <version>0.01-${project.parent.version}</version>
```
Comment (lloydmeta) [inline clandmark/pom.xml]: Could not figure out what version CLandmark is at, so left this at 0.01..

File: clandmark/src/main/java/org/bytedeco/javacpp/presets/clandmark.java [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+
+/**
+ * @author Lloyd @ github.com/lloydmeta
+ */
+@Properties(value = @Platform(
+        define = "_MSVC_COMPAT__",
+        include = {"CLandmark.h", "Flandmark.h", "CFeaturePool.h", "CFeatures.h"},
+        link = { "clandmark@.1.5", "flandmark@.1.5"}),
+        inherit = opencv_imgproc.class, target = "org.bytedeco.javacpp.clandmark")
+public class clandmark implements InfoMapper {
```
Comment (lloydmeta) [inline clandmark/src/main/java/org/bytedeco/javacpp/presets/clandmark.java]: Please take a look at the mappings and annotations here...

**Comment by lloydmeta** [comment]
CC @saudet  and @jpsacha

File: clandmark/src/main/java/org/bytedeco/javacpp/presets/clandmark.java [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+    public void map(InfoMap infoMap) {
+        infoMap.put(new Info("fl_double_t", "clandmark::fl_double_t").cast().valueTypes("float").pointerTypes("FloatPointer").define());
+        infoMap.put(new Info("cimg_library::CImg<unsigned char>").cast().pointerTypes("BytePointer"));
```
Comment (lloydmeta) [inline clandmark/src/main/java/org/bytedeco/javacpp/presets/clandmark.java]: This one in particular is dubious; "unsigned char" itself needs translation, and `CImg` is a generic type..

**Comment by lloydmeta** [comment]
After taking another look, I realised that CImg is not in fact what I wanted to cast it as; looks like I need to include `CImg.h` in order to have that class available, but then once I include it, the resulting file fails to compile with a lot of errors. Definitely need some more experienced eyes looking at this.

**Comment by saudet** [comment]
Right, the problem is, it's hard to figure out automatically which header files contain the actual interface we're interested in, and which ones contain implementation details and what not. To compensate, users need to include all the files they want to process, manually, in the correct order. So, check inside `CImg.h` to see which other files get included, and list those up as well, a potential recursive process... :)

**Comment by lloydmeta** [comment]
No, `CImg.h` is the only file in that library, albeit a 2.59MB one; see [the repo](https://github.com/dtschump/CImg). The project prides itself on being one single header file.

So, yeah, look for `struct CImg {` (line 9307) and you will see it's in that file, which subsequently includes a *lot* of stuff (the entire file is 46180 lines long), some of which the parser can't handle, and likely a lot that the parser isn't handling properly. Some more specific details on what to do in this situation would be helpful.

**Comment by saudet** [comment]
Ah, I see. The CImg library is separate from CLandmark though, but it's a requirement? Hum... It's a template library, so we'll need to create instances manually, as shown for Caffe here for example:
https://github.com/bytedeco/javacpp-presets/blob/master/caffe/src/main/java/org/bytedeco/javacpp/presets/caffe.java#L95
So, for example, we'd have `ByteCImg` for `CImg<unsigned char>` I guess.

**Comment by lloydmeta** [comment]
Ah ok, let me give that a try ! Thanks.

**Comment by saudet** [comment]
Yeah, it looks like it's going to be a hard one if we need to parse CImg as well :(

**Comment by saudet** [comment]
Maybe we could provide a small amount of accessor methods in a custom header file, and map those instead of all of CImg? Sounds a lot easier to do! Want to try this out?

**Comment by lloydmeta** [comment]
@saudet I've hard-pressed for time and so didn't really progress with this PR. I'll close it now to give someone else a try. Your idea sounds great: maybe you should try it out and see how far you get with it?

**Comment by saudet** [comment]
Or @jpsacha?

**Comment by saudet** [comment]
Actually, a better idea would be to provide a couple of functions to convert between CImg and OpenCV's Mat, so it would work pretty much like the current wrappers for flandmark from JavaCV's point of view. I think that'd be great and should be easy to do. Let me know when you guys have some time to work on that!

**Pull request closed without merging**